core/src/main/java/com/linecorp/armeria/client/ClientConnectionTimingsBuilder.java

```diff
@@ -0,0 +1,116 @@
+     * {@link ClientConnectionTimingsBuilder} is considered as the start time of resolving a domain name.
+     */
+    public ClientConnectionTimingsBuilder dnsResolutionEnd() {
+        // The start time of dnsResolution is acquiringConnectionStartMicros and acquiringConnectionStartNanos.
+        // So we don't have to call checkState() here.
+        dnsResolutionEndNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when connecting to a remote peer started.
+     */
+    public ClientConnectionTimingsBuilder socketConnectStart() {
+        socketConnectStartMicros = SystemInfo.currentTimeMicros();
+        socketConnectStartNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when connecting to a remote peer ends.
+     *
+     * @throws IllegalStateException if {@link #socketConnectStart()} is not invoked before calling this.
+     */
+    public ClientConnectionTimingsBuilder socketConnectEnd() {
+        checkState(socketConnectStartMicros >= 0, "socketConnectStart is not called yet.");
+        socketConnectEndNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when waiting an ongoing connecting attempt started in order to use one connection
+     * for HTTP/2.
+     */
+    public ClientConnectionTimingsBuilder pendingAcquisitionStart() {
+        pendingAcquisitionStartMicros = SystemInfo.currentTimeMicros();
+        pendingAcquisitionStartNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when waiting an ongoing connecting attempt ends in order to use one connection
+     * for HTTP/2.
+     *
+     * @throws IllegalStateException if {@link #pendingAcquisitionStart()} is not invoked before calling this.
+     */
+    public ClientConnectionTimingsBuilder pendingAcquisitionEnd() {
+        checkState(pendingAcquisitionStartMicros >= 0, "pendingAcquisitionStart is not called yet.");
+        pendingAcquisitionEndNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Returns a newly-created {@link ClientConnectionTimings} instance.
+     */
+    public ClientConnectionTimings build() {
+        return new ClientConnectionTimings(
+                acquiringConnectionStartMicros,
+                System.nanoTime() - acquiringConnectionStartNanos,
+                dnsResolutionEndNanos == 0 ? -1 : acquiringConnectionStartMicros,
+                dnsResolutionEndNanos == 0 ? -1 : dnsResolutionEndNanos - acquiringConnectionStartNanos,
+                socketConnectStartMicros == 0 ? -1 : socketConnectStartMicros,
+                socketConnectEndNanos == 0 ? -1 : socketConnectEndNanos - socketConnectStartNanos,
+                pendingAcquisitionStartMicros == 0 ? -1 : pendingAcquisitionStartMicros,
+                pendingAcquisitionEndNanos == 0 ? -1 : pendingAcquisitionEndNanos -
```

Comment:
There's a chance where `System.nanoTime()` returns `0`. I think we need `boolean` fields that tells if each property was set.